### PR TITLE
Tracking

### LINKED ISSUE #167: Tracking
We would consider 2 tracking options for scatter series, linear series, and bar series. The first one will be the closest existing point from all the series (this could be archived using a [Voronoi diagram](https://observablehq.com/@mbostock/the-delaunays-dual)).

The second option is to find the closest X value, considering X as the independent axis. This should return the closest value of all the series at the X-axis. The range series only should implement this case, returning the corresponding limits.

Will be needed to see the values of series when I move the mouse over a position in a plot, giving a [tooltip](https://airbnb.io/visx/docs/tooltip) or visual hint of the tracked position and the requested data, similar to this case https://airbnb.io/visx/xychart

example

### PR DESCRIPTION
**Pull request by maasencioh**
closes #167

### CODE REVIEW

File: src/Plot.tsx [hunk 1]
```diff
@@ -130,6 +134,29 @@ export default function Plot(props: PlotProps) {
               </g>
 
               {axes}
+
+              <Tracking
+                onClick={(position) => {
```
Comment (stropitek) [inline src/Plot.tsx:139]: Could pass undefined if callback is undefined

File: src/Plot.tsx [hunk 2]
```diff
@@ -130,6 +134,29 @@ export default function Plot(props: PlotProps) {
...
+                  onClick?.(
+                    closestCalculation(
+                      position,
+                      closest,
+                      state.series,
+                      axisContext,
+                    ),
+                  );
+                }}
+                onMouseMove={(position) => {
```
Comment (stropitek) [inline src/Plot.tsx:149]: ditto

**Review by stropitek — COMMENTED** [review]
I think you should add the `onMouseEnter` and `onMouseLeave` callbacks to toggle the display of the annotation when the mouse leaves the plot

File: src/Plot.tsx (hunk 1 above)
Comment (targos) [reply src/Plot.tsx:139]: More importantly, I think we should not render the <Tracking component at all if there are no callbacks, to fully optimize the common case.

File: src/types.ts [hunk 3]
```diff
@@ -83,6 +83,18 @@ export interface PlotProps {
...
+   */
+  onMouseMove?: (result: TrackingResult) => void;
+  /**
+   * Track values on mouse click
+   */
+  onClick?: (result: TrackingResult) => void;
+  /**
+   * Track closest values
+   */
+  closest?: { x?: boolean; y?: boolean; euclidean?: boolean };
```
[the remark this answers is not in the document]
Comment (stropitek) [reply src/types.ts:113]: https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIBsD2BnC2yqYgzADmyA3gLABQy9yEArgusACYRwgD8AXMgBGmTOi4gA3LQbIAHvyEix3KXQYBPBcNHjVAX1q1QkWIhSFMUdqDiRslaQ1kCQTALaDoqmepfvPUPqGdGDqAA7mWLj4AGpw6Ex4ADwAKoyykCDs9hg4eAREJKQAfMgAvA5q9MysHOICKQDaAOQ1bJzczQC66ZnZyGBQicg8aCJWNnbIAkxZECQgEOzeTg0tst29EFn2g8OjFhMgtnjTyLOcC0sr9L7ITc3qmxAZ2-17KAfj1sdTM3NXZa0Aw0WgwWYIMDAIhjNxhIjbMCpLY7NBRfKEYhkYoACjCUEwYWwAmoVWQrDy+AaN2QRAwwAQAGsBDiAG7xRICXLRMBxBLJFLFACU5VKrMwHH0ItJMigEDATCgIGQrnQ6CCoJohDhCPAOJlDApPJJjhk1RY7XqAyGEAANKbkHp7WS6WwmSz2egRWVSgazcgAPQB5CXUBLB0yBBEfCMC11bjlZCegB0bXjkgjDCDKswYBD8zD7Ez9CjIBjskTKdkNJBeiFwSAA

[the remark this answers is not in the document]
Comment (targos) [reply src/types.ts:113]: What if we need multiple closest values at the same time?

File: stories/control/tracking.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+  return (
+    <div>
+      <Plot
+        {...plot}
+        margin={{ bottom: 45, left: 90, top: 40, right: 20 }}
+        onMouseMove={({ coordinates, event: { clientX, clientY } }) =>
+          setHover({ coordinates, position: { x: clientX, y: clientY } })
+        }
+        onClick={({ series }) => setClosest(series)}
```
Comment (targos) [inline stories/control/tracking.stories.tsx:46]: result => result.getClosest('euclidean')

File: src/types.ts [hunk 5]
```diff
@@ -175,6 +195,25 @@ export interface MarkersProps {
...
+  point: SeriesPointType;
+  axis: AxisContextType[];
+}
+export interface TrackingResult {
+  event: React.MouseEvent<SVGRectElement, MouseEvent>;
+  coordinates: Record<string, number>;
+  series?: Record<
+    string,
+    { closestX?: ClosestInfo; closestY?: ClosestInfo; closest?: ClosestInfo }
+  >;
```
Comment (targos) [inline src/types.ts:208]:
```suggestion
  getClosest('x' | 'y' | 'euclidean'): Record<string, ClosestInfo>;
```

**Review by targos — APPROVED** [review]

**Pull request merged**